File: audit/src/lib/attestation/service.ts
import { createHash, randomBytes } from "crypto";
import { prisma } from "@/lib/db";

export interface AttestationInput {
  auditCaseId: string;
  creativeHash: string;
  destinationUrl: string;
  placementDomains: string[];
  policyVersion: string;
  expiresInDays?: number;
}

export interface AttestationOutput {
  attestationId: string;
  creativeHash: string;
  destinationHash: string;
  placementDomainHash: string;
  policyVersionHash: string;
  issuedAt: number;
  expiresAt: number;
}

function sha256(data: string): string {
  return "0x" + createHash("sha256").update(data).digest("hex");
}

export async function issueAttestation(
  input: AttestationInput
): Promise<AttestationOutput> {
  const attestationId =
    "0x" + randomBytes(32).toString("hex");
  const now = Math.floor(Date.now() / 1000);
  const expiresAt =
    now + (input.expiresInDays ?? 30) * 24 * 60 * 60;

  const destinationHash = sha256(input.destinationUrl);
  const placementDomainHash = sha256(
    input.placementDomains.sort().join(",")
  );
  const policyVersionHash = sha256(input.policyVersion);

  // Save to database
  await prisma.attestation.create({
    data: {
      auditCaseId: input.auditCaseId,
      attestationId,
      chainId: 84532, // Base Sepolia
      status: "ACTIVE",
      issuedAt: new Date(now * 1000),
      expiresAt: new Date(expiresAt * 1000),
    },
  });

  // TODO: In production, call the on-chain contract here
  // const tx = await registryContract.issueAttestation(...)

  return {
    attestationId,
    creativeHash: input.creativeHash,
    destinationHash,
    placementDomainHash,
    policyVersionHash,
    issuedAt: now,
    expiresAt,
  };
}

export async function generateManifest(
  creativeId: string,
  attestation: AttestationOutput,
  creative: {
    projectName: string;
    imageUrl: string;
    clickUrl: string | null;
    landingUrl: string;
  }
) {
  const manifestJson = {
    manifestId: `mf_${randomBytes(8).toString("hex")}`,
    creativeId,
    projectName: creative.projectName,
    creativeUrl: creative.imageUrl,
    clickUrl: creative.clickUrl || creative.landingUrl,
    declaredLandingUrl: creative.landingUrl,
    chainId: 84532,
    registryAddress: process.env.REGISTRY_ADDRESS || "0x0000000000000000000000000000000000000000",
    attestationId: attestation.attestationId,
    creativeHash: attestation.creativeHash,
    destinationHash: attestation.destinationHash,
    policyVersion: "v1.0",
    issuedAt: attestation.issuedAt,
    expiresAt: attestation.expiresAt,
    issuer: process.env.ISSUER_ADDRESS || "0x0000000000000000000000000000000000000000",
    reportUrl: `/api/reports/${attestation.attestationId}`,
  };

  const manifest = await prisma.manifest.create({
    data: {
      creativeId,
      attestationId: attestation.attestationId,
      manifestJson,
    },
  });

  return manifest;
}

File: audit/backend/internal/attestation/service.go
package attestation

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"sort"
	"strings"
	"time"
)

// AttestationInput contains the parameters needed to issue an attestation.
type AttestationInput struct {
	AuditCaseID      string
	CreativeHash     string
	DestinationURL   string
	PolicyVersion    string
	PlacementDomains []string
	ExpiresInDays    int
}

// AttestationOutput is the result of issuing an attestation.
type AttestationOutput struct {
	AttestationID       string
	CreativeHash        string
	DestinationHash     string
	PlacementDomainHash string
	PolicyVersionHash   string
	IssuedAt            int64
	ExpiresAt           int64
}

// CreativeInfo holds creative metadata used when generating a manifest.
type CreativeInfo struct {
	ProjectName string
	ImageURL    string
	ClickURL    string
	LandingURL  string
}

// ManifestData represents the full manifest JSON structure.
type ManifestData struct {
	ManifestID         string `json:"manifestId"`
	CreativeID         string `json:"creativeId"`
	ProjectName        string `json:"projectName"`
	CreativeURL        string `json:"creativeUrl"`
	ClickURL           string `json:"clickUrl"`
	DeclaredLandingURL string `json:"declaredLandingUrl"`
	ChainID            int    `json:"chainId"`
	RegistryAddress    string `json:"registryAddress"`
	AttestationID      string `json:"attestationId"`
	CreativeHash       string `json:"creativeHash"`
	DestinationHash    string `json:"destinationHash"`
	PolicyVersion      string `json:"policyVersion"`
	IssuedAt           int64  `json:"issuedAt"`
	ExpiresAt          int64  `json:"expiresAt"`
	Issuer             string `json:"issuer"`
	ReportURL          string `json:"reportUrl"`
}

// sha256Hex computes the SHA-256 hash of data and returns it as a 0x-prefixed hex string.
func sha256Hex(data string) string {
	h := sha256.Sum256([]byte(data))
	return "0x" + hex.EncodeToString(h[:])
}

// randomHex generates n random bytes and returns them as a hex string.
func randomHex(n int) (string, error) {
	b := make([]byte, n)
	if _, err := rand.Read(b); err != nil {
		return "", fmt.Errorf("failed to generate random bytes: %w", err)
	}
	return hex.EncodeToString(b), nil
}

// IssueAttestation creates a new attestation from the given input.
func IssueAttestation(input AttestationInput) (AttestationOutput, error) {
	// Generate random 32-byte attestation ID
	randID, err := randomHex(32)
	if err != nil {
		return AttestationOutput{}, err
	}
	attestationID := "0x" + randID

	now := time.Now().Unix()

	expiresInDays := input.ExpiresInDays
	if expiresInDays <= 0 {
		expiresInDays = 30
	}
	expiresAt := now + int64(expiresInDays)*24*60*60

	// Compute hashes
	destinationHash := sha256Hex(input.DestinationURL)

	// Sort placement domains and join with comma, matching the TS implementation
	domains := make([]string, len(input.PlacementDomains))
	copy(domains, input.PlacementDomains)
	sort.Strings(domains)
	placementDomainHash := sha256Hex(strings.Join(domains, ","))

	policyVersionHash := sha256Hex(input.PolicyVersion)

	return AttestationOutput{
		AttestationID:       attestationID,
		CreativeHash:        input.CreativeHash,
		DestinationHash:     destinationHash,
		PlacementDomainHash: placementDomainHash,
		PolicyVersionHash:   policyVersionHash,
		IssuedAt:            now,
		ExpiresAt:           expiresAt,
	}, nil
}

// GenerateManifest creates a ManifestData struct from an attestation and creative info.
func GenerateManifest(creativeID string, att AttestationOutput, creative CreativeInfo, registryAddress string, issuerAddress string) (ManifestData, error) {
	randID, err := randomHex(8)
	if err != nil {
		return ManifestData{}, err
	}
	manifestID := "mf_" + randID

	clickURL := creative.ClickURL
	if clickURL == "" {
		clickURL = creative.LandingURL
	}

	if registryAddress == "" {
		registryAddress = "0x0000000000000000000000000000000000000000"
	}
	if issuerAddress == "" {
		issuerAddress = "0x0000000000000000000000000000000000000000"
	}

	return ManifestData{
		ManifestID:         manifestID,
		CreativeID:         creativeID,
		ProjectName:        creative.ProjectName,
		CreativeURL:        creative.ImageURL,
		ClickURL:           clickURL,
		DeclaredLandingURL: creative.LandingURL,
		ChainID:            84532, // Base Sepolia
		RegistryAddress:    registryAddress,
		AttestationID:      att.AttestationID,
		CreativeHash:       att.CreativeHash,
		DestinationHash:    att.DestinationHash,
		PolicyVersion:      "v1.0",
		IssuedAt:           att.IssuedAt,
		ExpiresAt:          att.ExpiresAt,
		Issuer:             issuerAddress,
		ReportURL:          fmt.Sprintf("/api/reports/%s", att.AttestationID),
	}, nil
}
